File: python/backuptool/dbdump.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""Database dump adapters.

Detect locally installed databases and dump them into the backup set's
``.backuptool-db/`` folder. Native adapters (run as root via the local socket):
PostgreSQL, MySQL/MariaDB, Redis, MongoDB. Anything else (Oracle, MS SQL, ...)
is covered by a *generic* adapter: the user supplies a dump command; it runs
with ``BACKUPTOOL_DB_OUT`` pointing at the output folder.
"""
from __future__ import annotations

import os
import shutil
import socket as _socket
import subprocess
import sys

DB_DIR = ".backuptool-db"

_PORTS = {"postgresql": 5432, "mysql": 3306, "redis": 6379, "mongodb": 27017}
_SOCKETS = {
    "postgresql": ["/var/run/postgresql/.s.PGSQL.5432", "/tmp/.s.PGSQL.5432"],
    "mysql": ["/run/mysqld/mysqld.sock", "/var/run/mysqld/mysqld.sock", "/tmp/mysql.sock"],
    "redis": ["/var/run/redis/redis-server.sock", "/run/redis/redis-server.sock"],
    "mongodb": ["/tmp/mongodb-27017.sock"],
}


def _has(binary: str) -> bool:
    return shutil.which(binary) is not None


def _am_root() -> bool:
    return hasattr(os, "geteuid") and os.geteuid() == 0


def is_running(kind: str, conn: dict = None) -> bool:
    """Best-effort check that the database is actually accepting connections
    (TCP port probe, then default/explicit unix-socket file)."""
    conn = conn or {}
    host = conn.get("host") or "127.0.0.1"
    port = conn.get("port") or _PORTS.get(kind)
    if port:
        try:
            with _socket.create_connection((host, int(port)), timeout=0.4):
                return True
        except OSError:
            pass
    socks = list(_SOCKETS.get(kind, []))
    if conn.get("socket"):
        socks.insert(0, conn["socket"])
    return any(os.path.exists(s) for s in socks)


def detect_databases() -> list:
    """Installed databases we can dump. Each: {"name", "kind", "running"}."""
    out = []
    if not sys.platform.startswith(("linux", "darwin")):
        return out
    if _has("pg_dumpall"):
        out.append({"name": "postgresql", "kind": "postgresql"})
    if _has("mysqldump") or _has("mariadb-dump"):
        out.append({"name": "mysql", "kind": "mysql"})
    if _has("redis-cli"):
        out.append({"name": "redis", "kind": "redis"})
    if _has("mongodump"):
        out.append({"name": "mongodb", "kind": "mongodb"})
    for d in out:
        d["running"] = is_running(d["kind"])
    return out


def _native_command(kind: str, out_dir: str, conn: dict = None):
    """Return (argv, stdout_path_or_None) for a built-in adapter, applying the
    optional connection settings (host/port/socket/user)."""
    conn = conn or {}
    host, port = conn.get("host"), conn.get("port")
    user, sock, pw = conn.get("user"), conn.get("socket"), conn.get("password")
    if kind == "postgresql":
        runner = (["runuser", "-u", "postgres", "--"] if _am_root()
                  else ["sudo", "-n", "-u", "postgres"] if _has("sudo") else [])
        cmd = runner + ["pg_dumpall"]
        if host: cmd += ["-h", host]
        if port: cmd += ["-p", str(port)]
        if user: cmd += ["-U", user]
        return cmd, os.path.join(out_dir, "postgresql-all.sql")
    if kind == "mysql":
        dump = "mariadb-dump" if _has("mariadb-dump") else "mysqldump"
        cmd = [dump, "--all-databases", "--single-transaction", "--routines", "--events"]
        if host: cmd += ["-h", host]
        if port: cmd += ["-P", str(port)]
        if user: cmd += ["-u", user]
        if sock: cmd += ["--socket", sock]
        return cmd, os.path.join(out_dir, "mysql-all.sql")
    if kind == "redis":
        cmd = ["redis-cli"]
        if host: cmd += ["-h", host]
        if port: cmd += ["-p", str(port)]
        if sock: cmd += ["-s", sock]
        if pw: cmd += ["-a", pw]
        cmd += ["--rdb", os.path.join(out_dir, "redis-dump.rdb")]
        return cmd, None
    if kind == "mongodb":
        cmd = ["mongodump"]
        if host: cmd += ["--host", host]
        if port: cmd += ["--port", str(port)]
        if user: cmd += ["-u", user]
        if pw: cmd += ["-p", pw]
        cmd += ["--out", os.path.join(out_dir, "mongodb")]
        return cmd, None
    return None, None


def dump_database(spec: dict, out_dir: str, log=print, conn: dict = None) -> bool:
    """Dump one database. ``spec`` is either {"kind": <builtin>} or
    {"name", "shell": <command>} for a generic/custom dump (e.g. Oracle).
    ``conn`` carries optional host/port/socket/user/password."""
    os.makedirs(out_dir, exist_ok=True)
    name = spec.get("name", spec.get("kind", "db"))

    if spec.get("shell"):
        outfile = os.path.join(out_dir, f"{name}.log")
        env = dict(os.environ, BACKUPTOOL_DB_OUT=out_dir)
        try:
            with open(outfile, "w", encoding="utf-8") as f:
                rc = subprocess.run(spec["shell"], shell=True, env=env,
                                    stdout=f, stderr=subprocess.STDOUT).returncode
        except OSError as e:
            log(f"  DB {name}: error {e}")
            return False
        log(f"  DB {name}: {'ok' if rc == 0 else f'FAILED (exit {rc})'}")
        return rc == 0

    argv, stdout_path = _native_command(spec["kind"], out_dir, conn)
    if not argv:
        log(f"  DB {name}: no adapter")
        return False
    env = dict(os.environ)
    pw = (conn or {}).get("password")
    if pw:  # safe password passing for the tools that read it from the env
        env.setdefault("PGPASSWORD", pw)
        env.setdefault("MYSQL_PWD", pw)
    try:
        if stdout_path:
            with open(stdout_path, "w", encoding="utf-8") as f:
                p = subprocess.run(argv, stdout=f, stderr=subprocess.PIPE, env=env)
        else:
            p = subprocess.run(argv, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE, env=env)
    except OSError as e:
        log(f"  DB {name}: error {e}")
        return False
    if p.returncode == 0:
        log(f"  DB {name}: ok")
        return True
    err = p.stderr.decode("utf-8", "replace").strip().splitlines()
    log(f"  DB {name}: FAILED — {err[-1] if err else 'exit ' + str(p.returncode)}")
    return False


def dump_all(specs: list, out_dir: str, log=print, conn: dict = None) -> int:
    """Dump every spec into out_dir; return the number that succeeded."""
    ok = 0
    for spec in specs:
        if dump_database(spec, out_dir, log, conn):
            ok += 1
    return ok

File: rust/src/dbdump.rs
// SPDX-License-Identifier: GPL-3.0-or-later
//! Database dump adapters. Detect locally installed databases and dump them into
//! the backup set's `.backuptool-db/` folder. Native adapters (run as root via
//! the local socket): PostgreSQL, MySQL/MariaDB, Redis, MongoDB. Anything else
//! (Oracle, MS SQL, ...) uses a *generic* adapter: a user-supplied shell command
//! that runs with `BACKUPTOOL_DB_OUT` pointing at the output folder.

use std::net::ToSocketAddrs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::Duration;

pub const DB_DIR: &str = ".backuptool-db";

#[derive(Clone, Debug)]
pub struct DbSpec {
    pub name: String,
    pub kind: String,          // postgresql|mysql|redis|mongodb|generic
    pub shell: Option<String>, // generic command (Oracle, MS SQL, ...)
    pub running: bool,
}

/// Optional connection settings applied to the native dump commands.
#[derive(Clone, Default, Debug)]
pub struct DbConn {
    pub host: Option<String>,
    pub port: Option<String>,
    pub user: Option<String>,
    pub password: Option<String>,
    pub socket: Option<String>,
}

fn default_port(kind: &str) -> Option<u16> {
    match kind {
        "postgresql" => Some(5432),
        "mysql" => Some(3306),
        "redis" => Some(6379),
        "mongodb" => Some(27017),
        _ => None,
    }
}

fn default_sockets(kind: &str) -> &'static [&'static str] {
    match kind {
        "postgresql" => &["/var/run/postgresql/.s.PGSQL.5432", "/tmp/.s.PGSQL.5432"],
        "mysql" => &["/run/mysqld/mysqld.sock", "/var/run/mysqld/mysqld.sock", "/tmp/mysql.sock"],
        "redis" => &["/var/run/redis/redis-server.sock", "/run/redis/redis-server.sock"],
        "mongodb" => &["/tmp/mongodb-27017.sock"],
        _ => &[],
    }
}

/// Best-effort check that the database accepts connections (TCP probe, then a
/// default/explicit unix socket file).
pub fn is_running(kind: &str, conn: &DbConn) -> bool {
    let host = conn.host.clone().unwrap_or_else(|| "127.0.0.1".into());
    let port = conn.port.as_ref().and_then(|p| p.parse::<u16>().ok()).or_else(|| default_port(kind));
    if let Some(p) = port {
        if let Ok(addrs) = format!("{host}:{p}").to_socket_addrs() {
            for a in addrs {
                if std::net::TcpStream::connect_timeout(&a, Duration::from_millis(400)).is_ok() {
                    return true;
                }
            }
        }
    }
    let mut socks: Vec<String> = default_sockets(kind).iter().map(|s| s.to_string()).collect();
    if let Some(s) = &conn.socket {
        socks.insert(0, s.clone());
    }
    socks.iter().any(|s| Path::new(s).exists())
}

fn has(bin: &str) -> bool {
    std::env::var_os("PATH")
        .map(|paths| std::env::split_paths(&paths).any(|p| p.join(bin).is_file()))
        .unwrap_or(false)
}

#[cfg(unix)]
fn am_root() -> bool {
    unsafe { libc::geteuid() == 0 }
}
#[cfg(not(unix))]
fn am_root() -> bool {
    false
}

pub fn detect_databases() -> Vec<DbSpec> {
    let mk = |name: &str, kind: &str| DbSpec { name: name.into(), kind: kind.into(), shell: None, running: false };
    let mut v = Vec::new();
    if has("pg_dumpall") {
        v.push(mk("postgresql", "postgresql"));
    }
    if has("mysqldump") || has("mariadb-dump") {
        v.push(mk("mysql", "mysql"));
    }
    if has("redis-cli") {
        v.push(mk("redis", "redis"));
    }
    if has("mongodump") {
        v.push(mk("mongodb", "mongodb"));
    }
    for d in v.iter_mut() {
        d.running = is_running(&d.kind, &DbConn::default());
    }
    v
}

fn native(kind: &str, out_dir: &Path, conn: &DbConn) -> Option<(Vec<String>, Option<PathBuf>)> {
    let (host, port, user, sock, pw) = (
        conn.host.as_deref(), conn.port.as_deref(), conn.user.as_deref(),
        conn.socket.as_deref(), conn.password.as_deref(),
    );
    let push = |argv: &mut Vec<String>, flag: &str, val: Option<&str>| {
        if let Some(v) = val {
            argv.push(flag.into());
            argv.push(v.into());
        }
    };
    match kind {
        "postgresql" => {
            let mut argv: Vec<String> = if am_root() {
                vec!["runuser".into(), "-u".into(), "postgres".into(), "--".into()]
            } else if has("sudo") {
                vec!["sudo".into(), "-n".into(), "-u".into(), "postgres".into()]
            } else {
                vec![]
            };
            argv.push("pg_dumpall".into());
            push(&mut argv, "-h", host);
            push(&mut argv, "-p", port);
            push(&mut argv, "-U", user);
            Some((argv, Some(out_dir.join("postgresql-all.sql"))))
        }
        "mysql" => {
            let dump = if has("mariadb-dump") { "mariadb-dump" } else { "mysqldump" };
            let mut argv = vec![dump.to_string(), "--all-databases".into(),
                                "--single-transaction".into(), "--routines".into(), "--events".into()];
            push(&mut argv, "-h", host);
            push(&mut argv, "-P", port);
            push(&mut argv, "-u", user);
            push(&mut argv, "--socket", sock);
            Some((argv, Some(out_dir.join("mysql-all.sql"))))
        }
        "redis" => {
            let mut argv = vec!["redis-cli".to_string()];
            push(&mut argv, "-h", host);
            push(&mut argv, "-p", port);
            push(&mut argv, "-s", sock);
            push(&mut argv, "-a", pw);
            argv.push("--rdb".into());
            argv.push(out_dir.join("redis-dump.rdb").to_string_lossy().into_owned());
            Some((argv, None))
        }
        "mongodb" => {
            let mut argv = vec!["mongodump".to_string()];
            push(&mut argv, "--host", host);
            push(&mut argv, "--port", port);
            push(&mut argv, "-u", user);
            push(&mut argv, "-p", pw);
            argv.push("--out".into());
            argv.push(out_dir.join("mongodb").to_string_lossy().into_owned());
            Some((argv, None))
        }
        _ => None,
    }
}

pub fn dump_database<L: Fn(&str)>(spec: &DbSpec, out_dir: &Path, conn: &DbConn, log: &L) -> bool {
    let _ = std::fs::create_dir_all(out_dir);
    let err_path = out_dir.join(format!("{}.err", spec.name));
    let err_file = || std::fs::File::create(&err_path).ok();

    let status = if let Some(cmd) = &spec.shell {
        let mut c = Command::new("sh");
        c.arg("-c").arg(cmd).env("BACKUPTOOL_DB_OUT", out_dir);
        if let Ok(f) = std::fs::File::create(out_dir.join(format!("{}.log", spec.name))) {
            c.stdout(Stdio::from(f));
        }
        if let Some(f) = err_file() {
            c.stderr(Stdio::from(f));
        }
        c.status()
    } else {
        let (argv, stdout_path) = match native(&spec.kind, out_dir, conn) {
            Some(x) => x,
            None => {
                log(&format!("  DB {}: no adapter", spec.name));
                return false;
            }
        };
        let mut c = Command::new(&argv[0]);
        c.args(&argv[1..]);
        if let Some(pw) = &conn.password {   // safe password passing for pg/mysql
            c.env("PGPASSWORD", pw);
            c.env("MYSQL_PWD", pw);
        }
        match &stdout_path {
            Some(p) => {
                if let Ok(f) = std::fs::File::create(p) {
                    c.stdout(Stdio::from(f));
                }
            }
            None => {
                c.stdout(Stdio::null());
            }
        }
        if let Some(f) = err_file() {
            c.stderr(Stdio::from(f));
        }
        c.status()
    };

    match status {
        Ok(s) if s.success() => {
            let _ = std::fs::remove_file(&err_path);
            log(&format!("  DB {}: ok", spec.name));
            true
        }
        Ok(s) => {
            log(&format!("  DB {}: FAILED ({s}); see {}", spec.name, err_path.display()));
            false
        }
        Err(e) => {
            log(&format!("  DB {}: error {e}", spec.name));
            false
        }
    }
}

pub fn dump_all<L: Fn(&str)>(specs: &[DbSpec], out_dir: &Path, conn: &DbConn, log: &L) -> usize {
    specs.iter().filter(|s| dump_database(s, out_dir, conn, log)).count()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[cfg(unix)]  // the generic adapter runs a POSIX shell command (no sh/printf on Windows)
    #[test]
    fn generic_dump_writes_output() {
        let dir = std::env::temp_dir().join("bt-dbdump-test");
        let _ = std::fs::remove_dir_all(&dir);
        let spec = DbSpec {
            name: "demo".into(),
            kind: "generic".into(),
            shell: Some("printf data > \"$BACKUPTOOL_DB_OUT/demo.txt\"".into()),
            running: false,
        };
        let ok = dump_database(&spec, &dir, &DbConn::default(), &|_m: &str| {});
        assert!(ok);
        assert!(dir.join("demo.txt").is_file());
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[cfg(unix)]  // POSIX shell command; on Windows there is no sh to run it
    #[test]
    fn generic_dump_failure_reported() {
        let dir = std::env::temp_dir().join("bt-dbdump-fail");
        let spec = DbSpec { name: "boom".into(), kind: "generic".into(), shell: Some("exit 3".into()), running: false };
        assert!(!dump_database(&spec, &dir, &DbConn::default(), &|_m: &str| {}));
        let _ = std::fs::remove_dir_all(&dir);
    }
}
